### PR TITLE
Add typeless APIs to `Get` `Index` and `Bulk`

### PR DESCRIPTION
**Pull request by s1monw**
This commit adds `GET|PUT /{index}/_doc/{id}` and relatives to
index, bulk and get. This is a step towards removing types from elasticsearch.
This change adds a default `doc` type as the type for requests against typeless
APIs. In the future we might improve this by dyncamically detecting the type
in the mapping if there is one already.
This chance also doesn't cut over existing rest tests to ensure tests work for
our BWC tests. Once this is backported to 5.5 we can cut over rest tests to use
the typeless API as well as adding deprecation headers to the typed APIs.

Relates to #24592
Relates to #15613

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java [hunk 1]
```diff
@@ -44,6 +47,9 @@ public RestIndexAction(Settings settings, RestController controller) {
         CreateHandler createHandler = new CreateHandler(settings);
         controller.registerHandler(PUT, "/{index}/{type}/{id}/_create", createHandler);
         controller.registerHandler(POST, "/{index}/{type}/{id}/_create", createHandler);
+        // new typeless API added in 5.5
+        controller.registerHandler(POST, "/{index}_doc", new DocTypeRestHandler(settings, this::prepareRequest)); // auto id creation
```
Comment (clintongormley) [inline core/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java]: Missing `/` in `/{index}_doc`

File: core/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java [hunk 2]
```diff
@@ -44,6 +47,9 @@ public RestIndexAction(Settings settings, RestController controller) {
...
+        controller.registerHandler(PUT, "/{index}/_doc/{id}", new DocTypeRestHandler(settings, this::prepareRequest));
     }
```
Comment (clintongormley) [inline core/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java]: Aren't we missing `PUT|POST /{index}/_doc/{id}/_create`?

**Review by clintongormley — CHANGES_REQUESTED** [review]
looks good, but we're missing the rest handler and tests for the `_create`  endpoint I think?

File: core/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/action/document/RestIndexAction.java]: I missed to push a commit - this is incomplete

**Comment by s1monw** [comment]
> looks good, but we're missing the rest handler and tests for the _create endpoint I think?

what is this create thing? can't folks use a opt_type?

**Comment by clintongormley** [comment]
> what is this create thing? can't folks use a opt_type?

The `_create` end point has been there since 0.early.  using an `op_type` is an option, but requires custom handling in the clients, which is why we added the `create.json` rest spec:

https://github.com/elastic/elasticsearch/blob/master/rest-api-spec/src/main/resources/rest-api-spec/api/create.json

**Comment by s1monw** [comment]
clients.... zzzz.... :)

**Comment by s1monw** [comment]
@clintongormley I pushed a new commit

**Review by clintongormley — APPROVED** [review]
Thanks Simon.  LGTM

**Review by jpountz — APPROVED** [review]
LGTM!

File: core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 3]
```diff
@@ -48,7 +48,7 @@
 import static java.util.Collections.emptyMap;
 
 public class DocumentMapper implements ToXContent {
-
+    public static String DEFAULT_DOC_TYPE = "doc";
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java]: Probably worth javadoc.

File: rest-api-spec/src/main/resources/rest-api-spec/test/bulk/60_without_type.yml [hunk 4]
```diff
@@ -0,0 +1,42 @@
+---
+"Array of objects without type":
```
Comment (nik9000) [inline rest-api-spec/src/main/resources/rest-api-spec/test/bulk/60_without_type.yml]: I think this is missing a `skip`.

**Review by nik9000 — APPROVED** [review]

**Comment by s1monw** [comment]
closing unmerged see https://github.com/elastic/elasticsearch/issues/15613#issuecomment-303727982

**Pull request closed without merging**